File: backend/app/schemas/erp.py
"""ARCH47-S1:schemas — request and response models of the ERP posting API (app/api/v1/erp.py).

The console's frontend/src/types/erp.ts mirrors every response model field for
field (verify_arch47 W3 compares them).
"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, ConfigDict, Field


class _In(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)


# -- catalog -----------------------------------------------------------------------------


class PresetInfo(BaseModel):
    key: str
    label: str
    auth_modes: list[str]
    config_keys: list[str]
    objects: list[str]
    idempotency: Optional[str] = None
    probe_objects: list[str]


class FormatInfo(BaseModel):
    key: str
    label: str
    transports: list[str]
    objects: list[str]


class ErpCatalog(BaseModel):
    formats: list[FormatInfo]
    transports: list[str]
    presets: list[PresetInfo]
    ack_modes: dict[str, list[str]]
    auth_modes: dict[str, list[str]]
    object_kinds: list[str]
    object_labels: dict[str, str]
    source_kinds: list[str]
    source_labels: dict[str, str]
    states: list[str]
    state_labels: dict[str, str]
    transforms: list[str]
    header_paths: dict[str, str]
    line_paths: dict[str, str]
    language: str


# -- targets ------------------------------------------------------------------------------


class TargetCreate(_In):
    name: str = Field(min_length=1, max_length=120)
    format: str
    transport: str
    preset: str = "NONE"
    ack_mode: Optional[str] = None
    auth_mode: Optional[str] = None
    config: dict[str, Any] = Field(default_factory=dict)
    credential: Optional[dict[str, Any]] = None
    lookup_prefix: Optional[str] = Field(default=None, max_length=25)
    auto_post: bool = False
    auto_sources: list[str] = Field(default_factory=list)
    auto_objects: list[str] = Field(default_factory=list)
    max_attempts: Optional[int] = Field(default=None, ge=1, le=10)
    ack_timeout_hours: Optional[int] = Field(default=None, ge=1, le=720)


class TargetUpdate(_In):
    name: Optional[str] = Field(default=None, min_length=1, max_length=120)
    status: Optional[str] = None
    ack_mode: Optional[str] = None
    auth_mode: Optional[str] = None
    config: Optional[dict[str, Any]] = None
    auto_post: Optional[bool] = None
    auto_sources: Optional[list[str]] = None
    auto_objects: Optional[list[str]] = None
    max_attempts: Optional[int] = Field(default=None, ge=1, le=10)
    ack_timeout_hours: Optional[int] = Field(default=None, ge=1, le=720)


class CredentialSet(_In):
    credential: dict[str, Any]


class TargetRow(BaseModel):
    id: uuid.UUID
    name: str
    format: str
    transport: str
    preset: str
    ack_mode: str
    auth_mode: str
    status: str
    host: Optional[str] = None
    credential_set: bool
    credential_fingerprint: Optional[str] = None
    credential_updated_at: Optional[datetime] = None
    auto_post: bool
    auto_post_since: Optional[datetime] = None
    auto_sources: list[str]
    auto_objects: list[str]
    max_attempts: int
    ack_timeout_hours: int
    objects: list[str]
    lookup_prefix: Optional[str] = None
    postings: int
    open_exceptions: int
    created_at: datetime
    updated_at: datetime


class TargetList(BaseModel):
    items: list[TargetRow]


class MappingRow(BaseModel):
    id: uuid.UUID
    object_kind: str
    version: int
    status: str
    spec: dict[str, Any]
    spec_sha: str
    note: Optional[str] = None
    created_at: datetime
    retired_at: Optional[datetime] = None


class TargetDetail(BaseModel):
    target: TargetRow
    config: dict[str, Any]
    mappings: list[MappingRow]
    lookup_tables: list[str]


class TestResult(BaseModel):
    ok: bool
    kind: str
    message: str
    detail: dict[str, Any] = Field(default_factory=dict)


# -- mappings --------------------------------------------------------------------------------


class MappingSave(_In):
    spec: dict[str, Any]
    note: Optional[str] = Field(default=None, max_length=300)


class MappingVersions(BaseModel):
    target_id: uuid.UUID
    object_kind: str
    active: Optional[MappingRow] = None
    versions: list[MappingRow]
    default_spec: dict[str, Any]


class MappingCheck(BaseModel):
    valid: bool
    problems: list[str]


# -- lookup tables ------------------------------------------------------------------------------


class LookupCreate(_In):
    name: str = Field(min_length=1, max_length=64)
    description: Optional[str] = Field(default=None, max_length=300)
    entries: dict[str, Any] = Field(default_factory=dict)


class LookupUpdate(_In):
    description: Optional[str] = Field(default=None, max_length=300)
    entries: Optional[dict[str, Any]] = None
    merge: bool = False


class LookupRow(BaseModel):
    id: uuid.UUID
    name: str
    description: Optional[str] = None
    entries: dict[str, str]
    entry_count: int
    revision: int
    used_by: list[str]
    updated_at: datetime


class LookupList(BaseModel):
    items: list[LookupRow]


# -- postings ----------------------------------------------------------------------------------


class PostRequest(_In):
    target_id: uuid.UUID
    source_kind: str
    source_id: uuid.UUID
    object_kinds: list[str] = Field(min_length=1, max_length=5)


class PreviewRequest(_In):
    target_id: uuid.UUID
    source_kind: str
    source_id: uuid.UUID
    object_kind: str
    spec: Optional[dict[str, Any]] = None


class ReviewAction(_In):
    note: Optional[str] = Field(default=None, max_length=500)
    reference: Optional[str] = Field(default=None, max_length=200)


class AcknowledgeRequest(_In):
    accepted: Optional[bool] = None
    reference: Optional[str] = Field(default=None, max_length=200)
    reason: Optional[str] = Field(default=None, max_length=500)
    response_text: Optional[str] = Field(default=None, max_length=1_000_000)


class PostingRow(BaseModel):
    id: uuid.UUID
    target_id: uuid.UUID
    target_name: str
    target_format: str
    target_preset: str
    object_kind: str
    source_kind: str
    source_id: uuid.UUID
    work_item_id: Optional[uuid.UUID] = None
    work_item_filename: Optional[str] = None
    origin: str
    state: str
    document_number: Optional[str] = None
    amount: Optional[str] = None
    currency: Optional[str] = None
    attempts: int
    max_attempts: int
    next_attempt_at: Optional[datetime] = None
    external_id: Optional[str] = None
    last_error: Optional[str] = None
    mapping_version: Optional[int] = None
    rendered_filename: Optional[str] = None
    content_sha: Optional[str] = None
    delivered_at: Optional[datetime] = None
    acknowledged_at: Optional[datetime] = None
    reviewed_at: Optional[datetime] = None
    erased: bool
    created_at: datetime
    updated_at: datetime


class PostingList(BaseModel):
    items: list[PostingRow]
    total: int
    counts_by_state: dict[str, int]


class AttemptRow(BaseModel):
    seq: int
    kind: str
    outcome: str
    http_status: Optional[int] = None
    message: Optional[str] = None
    detail: dict[str, Any]
    actor_user_id: Optional[uuid.UUID] = None
    started_at: datetime


class PostingDetail(BaseModel):
    posting: PostingRow
    canonical: Optional[dict[str, Any]] = None
    mapped: Optional[dict[str, Any]] = None
    rendered_preview: Optional[str] = None
    rendered_media_type: Optional[str] = None
    rendered_size: Optional[int] = None
    ack: dict[str, Any]
    control_numbers: dict[str, Any]
    remote_path: Optional[str] = None
    idempotency_key: str
    source_label: Optional[str] = None
    source_changed: Optional[bool] = None
    attempts: list[AttemptRow]
    review_note: Optional[str] = None


class PostResult(BaseModel):
    object_kind: str
    posting_id: Optional[uuid.UUID] = None
    created: bool
    state: Optional[str] = None
    error: Optional[str] = None


class PostResponse(BaseModel):
    results: list[PostResult]


class PreviewResponse(BaseModel):
    ok: bool
    canonical: Optional[dict[str, Any]] = None
    mapped: Optional[dict[str, Any]] = None
    rendered_preview: Optional[str] = None
    rendered_media_type: Optional[str] = None
    filename: Optional[str] = None
    problems: list[str]
    notes: list[str]


class OutcomePosting(BaseModel):
    target_id: uuid.UUID
    object_kind: str
    posting_id: Optional[uuid.UUID] = None
    state: Optional[str] = None


class OutcomeRow(BaseModel):
    kind: str
    id: uuid.UUID
    label: str
    approved_at: Optional[datetime] = None
    work_item_id: Optional[uuid.UUID] = None
    documents: dict[str, str]
    objects: list[str]
    postings: list[OutcomePosting]


class OutcomeList(BaseModel):
    items: list[OutcomeRow]


__all__ = [name for name in dir() if name[0].isupper()]

File: frontend/src/types/erp.ts
/**
 * ARCH47-S2:types — ERP & System-of-Record Posting. Mirrors
 * backend/app/schemas/erp.py field for field (verify_arch47 W3 compares them)
 * and the vocabulary in backend/app/services/erp/vocabulary.py.
 */

export type ObjectKind = "VENDOR_BILL" | "PURCHASE_ORDER" | "GOODS_RECEIPT" | "JOURNAL_ENTRY" | "PAYMENT_REFERENCE";
export type SourceKind = "PROCUREMENT_CASE" | "TABLE" | "CASE";
export type PostingFormat = "CSV" | "XLSX" | "X12" | "UBL" | "TALLY" | "JSON";
export type PostingTransport = "DOWNLOAD" | "SFTP" | "HTTP";
export type PostingPreset =
  | "NONE" | "QUICKBOOKS_ONLINE" | "ZOHO_BOOKS" | "BUSINESS_CENTRAL" | "S4HANA_ODATA" | "NETSUITE_REST"
  | "GENERIC_REST" | "GENERIC_ODATA";
export type AckMode = "SYNC" | "X12_997" | "ACK_FILE" | "DELIVERY" | "MANUAL";
export type AuthMode =
  | "NONE" | "BEARER" | "BASIC" | "API_KEY_HEADER" | "OAUTH2_REFRESH_TOKEN" | "OAUTH2_CLIENT_CREDENTIALS"
  | "SSH_PASSWORD" | "SSH_KEY";
export type TargetStatus = "ACTIVE" | "DISABLED";
export type MappingStatus = "ACTIVE" | "RETIRED";
export type PostingState =
  | "PENDING" | "SENDING" | "RETRYING" | "DELIVERED" | "DONE" | "FAILED" | "REJECTED" | "MISMATCH" | "UNCERTAIN"
  | "CANCELLED";
export type PostingOrigin = "MANUAL" | "AUTO" | "FLOW";
export type AttemptKind = "RENDER" | "SEND" | "PROBE" | "ACK" | "REVIEW";
export type AttemptOutcome =
  | "OK" | "TRANSIENT" | "PERMANENT" | "UNCERTAIN" | "FOUND" | "ABSENT" | "PENDING" | "ACCEPTED" | "REJECTED"
  | "MISMATCH";
export type PostingVerdict = "RETRY" | "ACCEPT" | "CANCEL";

export const OBJECT_KINDS: readonly ObjectKind[] = [
  "VENDOR_BILL", "PURCHASE_ORDER", "GOODS_RECEIPT", "JOURNAL_ENTRY", "PAYMENT_REFERENCE",
];
export const SOURCE_KINDS: readonly SourceKind[] = ["PROCUREMENT_CASE", "TABLE", "CASE"];
export const POSTING_STATES: readonly PostingState[] = [
  "PENDING", "SENDING", "RETRYING", "DELIVERED", "DONE", "FAILED", "REJECTED", "MISMATCH", "UNCERTAIN", "CANCELLED",
];
export const EXCEPTION_STATES: readonly PostingState[] = ["FAILED", "REJECTED", "MISMATCH", "UNCERTAIN"];
export const OPEN_STATES: readonly PostingState[] = ["PENDING", "SENDING", "RETRYING", "DELIVERED"];

export const OBJECT_LABELS: Readonly<Record<ObjectKind, string>> = {
  VENDOR_BILL: "Vendor bill",
  PURCHASE_ORDER: "Purchase order",
  GOODS_RECEIPT: "Goods receipt",
  JOURNAL_ENTRY: "Journal entry",
  PAYMENT_REFERENCE: "Payment reference",
};

export const SOURCE_LABELS: Readonly<Record<SourceKind, string>> = {
  PROCUREMENT_CASE: "Reconciled invoice",
  TABLE: "Confirmed table",
  CASE: "Completed case",
};

export const STATE_LABELS: Readonly<Record<PostingState, string>> = {
  PENDING: "Queued",
  SENDING: "Sending",
  RETRYING: "Retrying",
  DELIVERED: "Awaiting acknowledgement",
  DONE: "Posted",
  FAILED: "Failed",
  REJECTED: "Rejected",
  MISMATCH: "Acknowledgement mismatch",
  UNCERTAIN: "Outcome unknown",
  CANCELLED: "Cancelled",
};

export const STATE_TONES: Readonly<Record<PostingState, string>> = {
  PENDING: "bg-slate-100 text-slate-700 dark:bg-slate-800 dark:text-slate-300",
  SENDING: "bg-sky-100 text-sky-800 dark:bg-sky-900/40 dark:text-sky-300",
  RETRYING: "bg-amber-100 text-amber-800 dark:bg-amber-900/40 dark:text-amber-300",
  DELIVERED: "bg-indigo-100 text-indigo-800 dark:bg-indigo-900/40 dark:text-indigo-300",
  DONE: "bg-emerald-100 text-emerald-800 dark:bg-emerald-900/40 dark:text-emerald-300",
  FAILED: "bg-rose-100 text-rose-800 dark:bg-rose-900/40 dark:text-rose-300",
  REJECTED: "bg-rose-100 text-rose-800 dark:bg-rose-900/40 dark:text-rose-300",
  MISMATCH: "bg-orange-100 text-orange-800 dark:bg-orange-900/40 dark:text-orange-300",
  UNCERTAIN: "bg-fuchsia-100 text-fuchsia-800 dark:bg-fuchsia-900/40 dark:text-fuchsia-300",
  CANCELLED: "bg-slate-100 text-slate-500 dark:bg-slate-800 dark:text-slate-400",
};

export const FORMAT_LABELS: Readonly<Record<PostingFormat, string>> = {
  CSV: "CSV (RFC 4180)",
  XLSX: "Excel workbook (.xlsx)",
  X12: "EDI X12 004010 (810 / 850 / 856)",
  UBL: "UBL 2.1 XML",
  TALLY: "Tally Prime XML",
  JSON: "REST / OData JSON",
};

export const PRESET_LABELS: Readonly<Record<PostingPreset, string>> = {
  NONE: "None",
  QUICKBOOKS_ONLINE: "QuickBooks Online",
  ZOHO_BOOKS: "Zoho Books",
  BUSINESS_CENTRAL: "Dynamics 365 Business Central",
  S4HANA_ODATA: "SAP S/4HANA (OData V2)",
  NETSUITE_REST: "Oracle NetSuite (REST)",
  GENERIC_REST: "Generic REST",
  GENERIC_ODATA: "Generic OData V4",
};

export const TRANSPORT_LABELS: Readonly<Record<PostingTransport, string>> = {
  DOWNLOAD: "Download (import by hand)",
  SFTP: "SFTP upload",
  HTTP: "HTTPS API",
};

export const ACK_LABELS: Readonly<Record<AckMode, string>> = {
  SYNC: "The API's response",
  X12_997: "X12 997 functional acknowledgement",
  ACK_FILE: "Acknowledgement file (.ack / Tally response)",
  DELIVERY: "Delivery only (file verified on the server)",
  MANUAL: "A person confirms the import",
};

export const AUTH_LABELS: Readonly<Record<AuthMode, string>> = {
  NONE: "None",
  BEARER: "Bearer token",
  BASIC: "Basic (username + password)",
  API_KEY_HEADER: "API key header",
  OAUTH2_REFRESH_TOKEN: "OAuth 2.0 refresh token",
  OAUTH2_CLIENT_CREDENTIALS: "OAuth 2.0 client credentials",
  SSH_PASSWORD: "SSH password",
  SSH_KEY: "SSH private key",
};

/** The fields each auth mode's credential takes (backend/app/services/erp/service.py _CREDENTIAL_FIELDS). */
export const CREDENTIAL_FIELDS: Readonly<Record<AuthMode, { required: readonly string[]; optional: readonly string[] }>> = {
  NONE: { required: [], optional: [] },
  BEARER: { required: ["token"], optional: [] },
  BASIC: { required: ["username", "password"], optional: [] },
  API_KEY_HEADER: { required: ["value"], optional: ["header"] },
  OAUTH2_REFRESH_TOKEN: { required: ["client_id", "client_secret", "refresh_token"], optional: [] },
  OAUTH2_CLIENT_CREDENTIALS: { required: ["client_id", "client_secret"], optional: [] },
  SSH_PASSWORD: { required: ["username", "password"], optional: [] },
  SSH_KEY: { required: ["username", "private_key"], optional: ["passphrase"] },
};

export const SECRET_FIELDS: ReadonlySet<string> = new Set(["token", "password", "value", "client_secret", "refresh_token",
  "private_key", "passphrase"]);

// -- catalog ------------------------------------------------------------------------------

export interface PresetInfo {
  readonly key: PostingPreset;
  readonly label: string;
  readonly auth_modes: readonly AuthMode[];
  readonly config_keys: readonly string[];
  readonly objects: readonly ObjectKind[];
  readonly idempotency?: string | null;
  readonly probe_objects: readonly ObjectKind[];
}

export interface FormatInfo {
  readonly key: PostingFormat;
  readonly label: string;
  readonly transports: readonly PostingTransport[];
  readonly objects: readonly ObjectKind[];
}

export interface ErpCatalog {
  readonly formats: readonly FormatInfo[];
  readonly transports: readonly PostingTransport[];
  readonly presets: readonly PresetInfo[];
  readonly ack_modes: Readonly<Record<string, readonly AckMode[]>>;
  readonly auth_modes: Readonly<Record<string, readonly AuthMode[]>>;
  readonly object_kinds: readonly ObjectKind[];
  readonly object_labels: Readonly<Record<string, string>>;
  readonly source_kinds: readonly SourceKind[];
  readonly source_labels: Readonly<Record<string, string>>;
  readonly states: readonly PostingState[];
  readonly state_labels: Readonly<Record<string, string>>;
  readonly transforms: readonly string[];
  readonly header_paths: Readonly<Record<string, string>>;
  readonly line_paths: Readonly<Record<string, string>>;
  readonly language: string;
}

// -- targets --------------------------------------------------------------------------------

export interface TargetCreate {
  readonly name: string;
  readonly format: PostingFormat;
  readonly transport: PostingTransport;
  readonly preset?: PostingPreset;
  readonly ack_mode?: AckMode | null;
  readonly auth_mode?: AuthMode | null;
  readonly config?: Record<string, unknown>;
  readonly credential?: Record<string, string> | null;
  readonly lookup_prefix?: string | null;
  readonly auto_post?: boolean;
  readonly auto_sources?: readonly SourceKind[];
  readonly auto_objects?: readonly ObjectKind[];
  readonly max_attempts?: number | null;
  readonly ack_timeout_hours?: number | null;
}

export interface TargetUpdate {
  readonly name?: string | null;
  readonly status?: TargetStatus | null;
  readonly ack_mode?: AckMode | null;
  readonly auth_mode?: AuthMode | null;
  readonly config?: Record<string, unknown> | null;
  readonly auto_post?: boolean | null;
  readonly auto_sources?: readonly SourceKind[] | null;
  readonly auto_objects?: readonly ObjectKind[] | null;
  readonly max_attempts?: number | null;
  readonly ack_timeout_hours?: number | null;
}

export interface CredentialSet {
  readonly credential: Record<string, string>;
}

export interface TargetRow {
  readonly id: string;
  readonly name: string;
  readonly format: PostingFormat;
  readonly transport: PostingTransport;
  readonly preset: PostingPreset;
  readonly ack_mode: AckMode;
  readonly auth_mode: AuthMode;
  readonly status: TargetStatus;
  readonly host?: string | null;
  readonly credential_set: boolean;
  readonly credential_fingerprint?: string | null;
  readonly credential_updated_at?: string | null;
  readonly auto_post: boolean;
  readonly auto_post_since?: string | null;
  readonly auto_sources: readonly SourceKind[];
  readonly auto_objects: readonly ObjectKind[];
  readonly max_attempts: number;
  readonly ack_timeout_hours: number;
  readonly objects: readonly ObjectKind[];
  readonly lookup_prefix?: string | null;
  readonly postings: number;
  readonly open_exceptions: number;
  readonly created_at: string;
  readonly updated_at: string;
}

export interface TargetList {
  readonly items: readonly TargetRow[];
}

export interface MappingRow {
  readonly id: string;
  readonly object_kind: ObjectKind;
  readonly version: number;
  readonly status: MappingStatus;
  readonly spec: Record<string, unknown>;
  readonly spec_sha: string;
  readonly note?: string | null;
  readonly created_at: string;
  readonly retired_at?: string | null;
}

export interface TargetDetail {
  readonly target: TargetRow;
  readonly config: Record<string, unknown>;
  readonly mappings: readonly MappingRow[];
  readonly lookup_tables: readonly string[];
}

export interface TestResult {
  readonly ok: boolean;
  readonly kind: string;
  readonly message: string;
  readonly detail: Record<string, unknown>;
}

// -- mappings ---------------------------------------------------------------------------------

export interface MappingSave {
  readonly spec: Record<string, unknown>;
  readonly note?: string | null;
}

export interface MappingVersions {
  readonly target_id: string;
  readonly object_kind: ObjectKind;
  readonly active?: MappingRow | null;
  readonly versions: readonly MappingRow[];
  readonly default_spec: Record<string, unknown>;
}

export interface MappingCheck {
  readonly valid: boolean;
  readonly problems: readonly string[];
}

// -- lookup tables ------------------------------------------------------------------------------

export interface LookupCreate {
  readonly name: string;
  readonly description?: string | null;
  readonly entries?: Record<string, string>;
}

export interface LookupUpdate {
  readonly description?: string | null;
  readonly entries?: Record<string, string> | null;
  readonly merge?: boolean;
}

export interface LookupRow {
  readonly id: string;
  readonly name: string;
  readonly description?: string | null;
  readonly entries: Readonly<Record<string, string>>;
  readonly entry_count: number;
  readonly revision: number;
  readonly used_by: readonly string[];
  readonly updated_at: string;
}

export interface LookupList {
  readonly items: readonly LookupRow[];
}

// -- postings -------------------------------------------------------------------------------------

export interface PostRequest {
  readonly target_id: string;
  readonly source_kind: SourceKind;
  readonly source_id: string;
  readonly object_kinds: readonly ObjectKind[];
}

export interface PreviewRequest {
  readonly target_id: string;
  readonly source_kind: SourceKind;
  readonly source_id: string;
  readonly object_kind: ObjectKind;
  readonly spec?: Record<string, unknown> | null;
}

export interface ReviewAction {
  readonly note?: string | null;
  readonly reference?: string | null;
}

export interface AcknowledgeRequest {
  readonly accepted?: boolean | null;
  readonly reference?: string | null;
  readonly reason?: string | null;
  readonly response_text?: string | null;
}

export interface PostingRow {
  readonly id: string;
  readonly target_id: string;
  readonly target_name: string;
  readonly target_format: PostingFormat;
  readonly target_preset: PostingPreset;
  readonly object_kind: ObjectKind;
  readonly source_kind: SourceKind;
  readonly source_id: string;
  readonly work_item_id?: string | null;
  readonly work_item_filename?: string | null;
  readonly origin: PostingOrigin;
  readonly state: PostingState;
  readonly document_number?: string | null;
  readonly amount?: string | null;
  readonly currency?: string | null;
  readonly attempts: number;
  readonly max_attempts: number;
  readonly next_attempt_at?: string | null;
  readonly external_id?: string | null;
  readonly last_error?: string | null;
  readonly mapping_version?: number | null;
  readonly rendered_filename?: string | null;
  readonly content_sha?: string | null;
  readonly delivered_at?: string | null;
  readonly acknowledged_at?: string | null;
  readonly reviewed_at?: string | null;
  readonly erased: boolean;
  readonly created_at: string;
  readonly updated_at: string;
}

export interface PostingList {
  readonly items: readonly PostingRow[];
  readonly total: number;
  readonly counts_by_state: Readonly<Record<string, number>>;
}

export interface AttemptRow {
  readonly seq: number;
  readonly kind: AttemptKind;
  readonly outcome: AttemptOutcome;
  readonly http_status?: number | null;
  readonly message?: string | null;
  readonly detail: Record<string, unknown>;
  readonly actor_user_id?: string | null;
  readonly started_at: string;
}

export interface PostingDetail {
  readonly posting: PostingRow;
  readonly canonical?: Record<string, unknown> | null;
  readonly mapped?: Record<string, unknown> | null;
  readonly rendered_preview?: string | null;
  readonly rendered_media_type?: string | null;
  readonly rendered_size?: number | null;
  readonly ack: Record<string, unknown>;
  readonly control_numbers: Record<string, unknown>;
  readonly remote_path?: string | null;
  readonly idempotency_key: string;
  readonly source_label?: string | null;
  readonly source_changed?: boolean | null;
  readonly attempts: readonly AttemptRow[];
  readonly review_note?: string | null;
}

export interface PostResult {
  readonly object_kind: ObjectKind;
  readonly posting_id?: string | null;
  readonly created: boolean;
  readonly state?: PostingState | null;
  readonly error?: string | null;
}

export interface PostResponse {
  readonly results: readonly PostResult[];
}

export interface PreviewResponse {
  readonly ok: boolean;
  readonly canonical?: Record<string, unknown> | null;
  readonly mapped?: Record<string, unknown> | null;
  readonly rendered_preview?: string | null;
  readonly rendered_media_type?: string | null;
  readonly filename?: string | null;
  readonly problems: readonly string[];
  readonly notes: readonly string[];
}

export interface OutcomePosting {
  readonly target_id: string;
  readonly object_kind: ObjectKind;
  readonly posting_id?: string | null;
  readonly state?: PostingState | null;
}

export interface OutcomeRow {
  readonly kind: SourceKind;
  readonly id: string;
  readonly label: string;
  readonly approved_at?: string | null;
  readonly work_item_id?: string | null;
  readonly documents: Readonly<Record<string, string>>;
  readonly objects: readonly ObjectKind[];
  readonly postings: readonly OutcomePosting[];
}

export interface OutcomeList {
  readonly items: readonly OutcomeRow[];
}

// -- helpers ----------------------------------------------------------------------------------------

export const isException = (state: PostingState): boolean => EXCEPTION_STATES.includes(state);

/** What a reviewer may decide for a posting in this state (mirrors service.review). */
export const verdictsFor = (state: PostingState): readonly PostingVerdict[] => {
  if (EXCEPTION_STATES.includes(state)) {
    return ["RETRY", "ACCEPT", "CANCEL"];
  }
  if (state === "RETRYING") {
    return ["RETRY", "CANCEL"];
  }
  if (state === "PENDING" || state === "DELIVERED") {
    return ["CANCEL"];
  }
  return [];
};

/** A starting configuration for a new target (the server validates every key). */
export const configTemplate = (format: PostingFormat, transport: PostingTransport, preset: PostingPreset): Record<string, unknown> => {
  const cfg: Record<string, unknown> = {};
  if (format === "CSV") {
    cfg.csv = { delimiter: "," };
  }
  if (format === "TALLY") {
    cfg.tally = { company: "" };
  }
  if (format === "X12") {
    cfg.x12 = { sender_qualifier: "ZZ", sender_id: "YOURSENDERID", receiver_qualifier: "ZZ", receiver_id: "PARTNERID", usage: "T" };
  }
  if (transport === "SFTP") {
    cfg.sftp = { host: "", port: 22, host_key_sha256: "SHA256:", directory: "/inbound" };
  }
  if (transport === "HTTP") {
    cfg.http = { base_url: "https://" };
    if (preset === "QUICKBOOKS_ONLINE") {
      cfg.qbo = { realm_id: "" };
    }
    if (preset === "ZOHO_BOOKS") {
      cfg.zoho = { organization_id: "" };
    }
    if (preset === "BUSINESS_CENTRAL") {
      cfg.bc = { company_id: "" };
    }
    if (preset === "S4HANA_ODATA") {
      cfg.s4 = { sap_client: "" };
    }
    if (preset === "GENERIC_REST" || preset === "GENERIC_ODATA") {
      cfg.rest = { paths: { VENDOR_BILL: "/bills" } };
    }
  }
  return cfg;
};
